lib/npm-auditer.js

```diff
@@ -3,114 +3,60 @@
+        throw new Error(
+          `Invocation of npm audit failed:\n${stderrBuffer.join('\n')}`
+        );
       }
-    });
-
-  if (whitelistedModulesFound.length) {
-    const found = whitelistedModulesFound.join(', ');
-    const msg = `Vulnerable whitelisted modules found: ${found}.`;
-    console.warn('\x1b[33m%s\x1b[0m', msg);
-  }
-  if (whitelistedAdvisoriesFound.length) {
-    const found = whitelistedAdvisoriesFound.join(', ');
-    const msg = `Vulnerable whitelisted advisories found: ${found}.`;
-    console.warn('\x1b[33m%s\x1b[0m', msg);
-  }
-
-  // Get the levels that have failed by filtering the keys with true values
-  const failedLevelsFound = Object.keys(failedLevels)
-    .filter(l => failedLevels[l])
-    .join(', ');
-  // If any of the levels have been failed
-  if (failedLevelsFound) {
-    // Get the levels that have failed by filtering the keys with true values
-    const err = `Failed security audit due to ${failedLevelsFound} vulnerabilities.`;
-    throw new Error(err);
-  }
-  return npmAudit;
-}
 
-function runNpmAudit(callback) {
-  childProcess.exec('npm audit --json', (_error, stdout, stderr) => {
-    if (stdout) {
+      const stdout = stdoutBuffer.join('\n');
       try {
-        const parsedAudit = JSON.parse(stdout);
-        callback(null, parsedAudit);
-        return;
-      } catch (error) {
-        callback(new Error('Invalid output'), null);
-        return;
+        return JSON.parse(stdout);
+      } catch (e) {
+        console.log(stdout);
+        throw e;
       }
-    }
-
-    if (_error) {
-      callback(new Error(stderr), null);
-    }
-  });
+    });
 }
 
 /**
  * Audit your NPM project!
  *
- * @param {{report: boolean, whitelist: string[], advisories: string[], levels: { low: boolean, moderate: boolean, high: boolean, critical: boolean }}} config
+ * @param {{dir: string, report: boolean, whitelist: string[], advisories: string[], levels: { low: boolean, moderate: boolean, high: boolean, critical: boolean }}} config
+ * `dir`: directory to run npm audit at.
  * `report`: whether to show the NPM audit report in the console.
  * `whitelist`: a list of packages that should not break the build if their vulnerability is found.
  * `advisories`: a list of advisory ids that should not break the build if found.
  * `levels`: the vulnerability levels to fail on, if `moderate` is set `true`, `high` and `critical` should be as well.
  * @returns {Promise<any>} Returns the audit report on resolve, `Error` on rejection.
  */
-function audit(config) {
-  return new Promise((resolve, reject) => {
-    runNpmAudit((err, result) => {
-      if (err) {
-        reject(err);
-        return;
-      }
-      try {
-        const report = reportAudit(result, config);
-        resolve(report);
-      } catch (error) {
-        reject(error);
-      }
-    });
-  });
+function audit(config, reporter = reportAudit) {
+  const model = new Model(config);
+
+  return Promise.resolve()
```

Comment:
Within this context, what benefit does `return Promise.resolve().then(() => runNpmAudit(config.dir))` used instead of `return runNpmAudit(config.dir)`? I came across [this StackOverflow link](https://stackoverflow.com/a/45113158) that describes why it's useful in some cases, but which case is it useful in this context?

Reply:
It due to the "Singular error handling" concern.

You mentioning it here turned my attention to the `new Model(config)` expression which can throw an exception so I moved it into the `then()` block to make sure that the only failures that can happen here are promise rejection (in other words: that this code cannot fail with a synchronously thrown exception).